cli/command/telemetry_utils.go

```diff
@@ -55,17 +54,27 @@ func (cli *DockerCli) InstrumentCobraCommands(cmd *cobra.Command, mp metric.Mete
 		}
 		cmd.RunE = func(cmd *cobra.Command, args []string) error {
 			// start the timer as the first step of every cobra command
-			baseAttrs := BaseCommandAttributes(cmd, cli)
-			stopCobraCmdTimer := startCobraCommandTimer(cmd, meter, baseAttrs)
+			stopInstrumentation := cli.StartInstrumentation(ctx, cmd)
 			cmdErr := ogRunE(cmd, args)
-			stopCobraCmdTimer(cmdErr)
+			stopInstrumentation(cmdErr)
 			return cmdErr
 		}
 
 		return ogPersistentPreRunE(cmd, args)
 	}
 }
 
+// StartInstrumentation instruments CLI commands with the individual metrics and spans configured.
+// It's the main command OTel utlility, and new command-related metrics should be added to it.
+// It should be called immediately before command execution, and returns a stopInstrumentation function
+// that must be called with the error resulting from the command execution.
+func (cli *DockerCli) StartInstrumentation(ctx context.Context, cmd *cobra.Command) (stopInstrumentation func(error)) {
+	mp := cli.MeterProvider(ctx)
```

Comment:
I'm a bit unsure if we should be creating a new MeterProvider here. In the case of the CLI itself, we could just get the Meter from the global MeterProvider (which would be already set by the time we invoke this function). In the case of plugins though, maybe we could do a check to see if a global MeterProvider has already been set and only create a new one if not set. WDYT?

Reply:
Sounds good to me, yeah!

Reply:
@jsternberg thoughts?

Reply:
Good observations @jsternberg 

> it might be good if we figure out who is responsible for initializing the providers

I'd say plugins should be responsible for initializing their own `Meter/TracerProviders`.  

IIUC this could be done in the `makeCmd` func passed into `plugin.Run()`, as long as it's done AFTER having run the `plugin.PersistentRunE()` func, so that the `dockerCli` struct is initialized.  

In the case of `buildx`, i guess it'd go here after calling `plugin.PersistentPreRunE()`: https://github.com/docker/buildx/blob/d4f088e689b41353d74f1a0bfcd6d7c0b213aed2/commands/root.go#L39

> buildx would need to be able to attach an additional span processor to the default provider because it has some special functionality to forward traces to buildkit for certain features to work

Can those `SpanProcessors` be attached to the `TracerProvider` as part of the `opts ...sdktrace.TracerProviderOption`, or am I missing something here? If so, buildx could just create it's `TracerProvider` using the CLI utils and set that as it's global `TracerProvider` 

I'm imagining something along the lines of this, done after `plugin.PersistentPreRunE()` as mentioned above:
```go
sp := newSpanProc()
tpOpts := []trace.TracerProviderOption{trace.WithSpanProcessor(sp)}
tp := dockerCli.TracerProvider(ctx, tpOpts)
otel.SetTracerProvider(tp)
```

> My fear is it might be too easy to use a provider retrieved from some global state in the wrong order and have no way to know whether it was done correctly than for someone to manually check.

This final part is a bit unclear to me atm, could you explain what exactly you mean by "wrong order" and how that affects things? I might be missing something here

Reply:
> This final part is a bit unclear to me atm, could you explain what exactly you mean by "wrong order" and how that affects things? I might be missing something here

Sorry my language was a bit unclear. If we were to use a global getter/setter within the plugins for retrieving the provider, you could easily end up in a situation like this:

```
// has this tracer provider been initialized already or is it pre-initialization?
tp := otel.GetTracerProvider()
```

I just want to avoid that question when accessing the tracer provider. If the cli framework is responsible for initializing the provider, then it would be easier to verify correctness if plugins ask the cli framework directly rather than relying on a side-effect.